File: backend/routes/attendance.js
const express = require('express');
const router = express.Router();
const multer = require('multer');
const path = require('path');
const fs = require('fs');
const { Class, Attendance } = require('../models/ClassAttendance');
const Student = require('../models/Student');
const auth = require('../middleware/auth');

// Setup multer for selfie uploads
const storage = multer.diskStorage({
  destination: (req, file, cb) => {
    const dir = 'uploads/selfies';
    if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
    cb(null, dir);
  },
  filename: (req, file, cb) => {
    cb(null, `selfie_${Date.now()}${path.extname(file.originalname)}`);
  },
});
const upload = multer({ storage, limits: { fileSize: 5 * 1024 * 1024 } });

// Haversine distance in meters
function getDistance(lat1, lon1, lat2, lon2) {
  const R = 6371000;
  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);
  const a = Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLon / 2) ** 2;
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}
function toRad(d) { return d * Math.PI / 180; }

// GET /api/attendance/check?studentId=&classId=
router.get('/check', auth, async (req, res) => {
  const { classId } = req.query;
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  const tomorrow = new Date(today); tomorrow.setDate(tomorrow.getDate() + 1);

  try {
    const record = await Attendance.findOne({
      student: req.student._id,
      class: classId,
      date: { $gte: today, $lt: tomorrow },
    });
    res.json({ signed: !!record });
  } catch (e) {
    res.status(500).json({ message: 'Server error' });
  }
});

// POST /api/attendance/sign
router.post('/sign', auth, upload.single('selfie'), async (req, res) => {
  const { classId, latitude, longitude, timestamp } = req.body;
  const RADIUS = parseInt(process.env.CLASSROOM_RADIUS_METERS) || 50;

  try {
    // 1. Find class
    const cls = await Class.findById(classId);
    if (!cls) return res.status(404).json({ message: 'Class not found.' });

    // 2. GPS Verification
    const dist = getDistance(
      parseFloat(latitude), parseFloat(longitude),
      cls.latitude, cls.longitude
    );
    if (dist > RADIUS) {
      return res.status(403).json({
        message: `You are ${Math.round(dist)}m from the classroom. Must be within ${RADIUS}m.`,
        distance: dist,
      });
    }

    // 3. Check duplicate
    const today = new Date(); today.setHours(0, 0, 0, 0);
    const tomorrow = new Date(today); tomorrow.setDate(tomorrow.getDate() + 1);
    const existing = await Attendance.findOne({
      student: req.student._id, class: classId,
      date: { $gte: today, $lt: tomorrow },
    });
    if (existing) return res.status(409).json({ message: 'Already signed attendance for this class today.' });

    // 4. Save attendance
    const selfieUrl = req.file ? `/uploads/selfies/${req.file.filename}` : null;
    const record = await Attendance.create({
      student: req.student._id,
      class: classId,
      date: new Date(),
      present: true,
      latitude: parseFloat(latitude),
      longitude: parseFloat(longitude),
      selfieUrl,
      verifiedByGPS: true,
      timestamp: timestamp || new Date(),
    });

    res.status(201).json({ success: true, record });
  } catch (e) {
    res.status(500).json({ message: 'Server error', error: e.message });
  }
});

// GET /api/attendance/stats/:studentId
router.get('/stats/:studentId', auth, async (req, res) => {
  try {
    const records = await Attendance.find({ student: req.student._id }).populate('class', 'name');
    const present = records.filter(r => r.present).length;
    const total = records.length;
    const absent = total - present;
    const rate = total > 0 ? Math.round((present / total) * 100) : 0;
    const needed = rate < 75 ? Math.ceil((0.75 * total - present) / 0.25) : 0;

    // Weekly data (last 7 days)
    const days = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
    const weeklyData = [0, 0, 0, 0, 0, 0, 0];
    const weeklyLabels = [];
    const now = new Date();
    for (let i = 6; i >= 0; i--) {
      const d = new Date(now); d.setDate(d.getDate() - i);
      weeklyLabels.push(days[d.getDay()]);
      const start = new Date(d); start.setHours(0, 0, 0, 0);
      const end = new Date(d); end.setHours(23, 59, 59, 999);
      const count = records.filter(r => r.present && r.date >= start && r.date <= end).length;
      weeklyData[6 - i] = count;
    }

    // Recent activity
    const recentActivity = records
      .sort((a, b) => b.date - a.date)
      .slice(0, 10)
      .map(r => ({
        className: r.class?.name || 'Unknown',
        date: r.date.toLocaleDateString(),
        present: r.present,
      }));

    res.json({ present, absent, total, rate, needed, weeklyData, weeklyLabels, recentActivity });
  } catch (e) {
    res.status(500).json({ message: 'Server error', error: e.message });
  }
});

// GET /api/attendance/reports/:studentId?period=month
router.get('/reports/:studentId', auth, async (req, res) => {
  const { period = 'month' } = req.query;
  const now = new Date();
  let startDate = new Date();

  if (period === 'week') startDate.setDate(now.getDate() - 7);
  else if (period === 'month') startDate.setMonth(now.getMonth() - 1);
  else if (period === 'semester') startDate.setMonth(now.getMonth() - 6);

  try {
    const classes = await Class.find({ enrolledStudents: req.student._id });
    const reports = await Promise.all(classes.map(async (cls) => {
      const records = await Attendance.find({
        student: req.student._id,
        class: cls._id,
        date: { $gte: startDate, $lte: now },
      });
      const present = records.filter(r => r.present).length;
      const total = records.length;
      const absent = total - present;
      const rate = total > 0 ? Math.round((present / total) * 100) : 0;
      const sessions = records.sort((a, b) => b.date - a.date).map(r => ({
        date: r.date.toLocaleDateString(),
        present: r.present,
      }));
      return { className: cls.name, courseCode: cls.courseCode, present, absent, total, rate, sessions };
    }));

    res.json(reports);
  } catch (e) {
    res.status(500).json({ message: 'Server error', error: e.message });
  }
});

module.exports = router;

File: backend/routes/attendance.py
from fastapi import APIRouter, HTTPException, Depends
from bson import ObjectId
from database import get_db
from models.schemas import AttendanceSign
from middleware.auth import get_current_user
from utils.geo import is_within_radius
from datetime import datetime, date

router = APIRouter()


def _fmt(doc):
    doc["_id"] = str(doc["_id"])
    return doc


@router.post("/sign")
async def sign_attendance(body: AttendanceSign, _=Depends(get_current_user)):
    db = get_db()

    # 1. Load class & verify it exists
    cls = await db.classes.find_one({"_id": ObjectId(body.class_id)})
    if not cls:
        raise HTTPException(404, "Class not found")

    # 2. Check student is enrolled
    if body.student_id not in cls.get("enrolled_students", []):
        raise HTTPException(403, "Not enrolled in this class")

    # 3. GPS geofence check (server-side)
    inside, distance_m = is_within_radius(
        body.latitude, body.longitude,
        cls["latitude"], cls["longitude"],
        cls.get("radius_meters", 50),
    )
    if not inside:
        raise HTTPException(400, f"You are {distance_m}m from the classroom (max {cls.get('radius_meters', 50)}m)")

    # 4. Duplicate prevention — one sign per student per class per day
    today_str = date.today().isoformat()
    existing = await db.attendance.find_one({
        "student_id": body.student_id,
        "class_id": body.class_id,
        "date": today_str,
    })
    if existing:
        raise HTTPException(409, "Already signed in for this class today")

    # 5. Record attendance
    record = {
        "student_id": body.student_id,
        "class_id": body.class_id,
        "class_name": cls["name"],
        "course_code": cls["course_code"],
        "date": today_str,
        "timestamp": datetime.utcnow(),
        "latitude": body.latitude,
        "longitude": body.longitude,
        "distance_m": distance_m,
        "selfie_base64": body.selfie_base64,
        "status": "present",
    }
    result = await db.attendance.insert_one(record)
    record["_id"] = str(result.inserted_id)
    record.pop("selfie_base64", None)  # don't echo it back
    return {"message": "Attendance recorded ✅", "record": record}


@router.get("/check")
async def check_today(student_id: str, class_id: str, _=Depends(get_current_user)):
    db = get_db()
    today_str = date.today().isoformat()
    rec = await db.attendance.find_one({
        "student_id": student_id,
        "class_id": class_id,
        "date": today_str,
    })
    return {"signed": rec is not None}


@router.get("/stats/{student_id}")
async def stats(student_id: str, _=Depends(get_current_user)):
    db = get_db()

    # Total enrolled classes
    enrolled_count = await db.classes.count_documents({"enrolled_students": student_id})

    # All attendance records
    records = await db.attendance.find({"student_id": student_id}).to_list(1000)
    total_sessions = len(records)

    # Per-class breakdown
    by_class: dict = {}
    for r in records:
        cid = r["class_id"]
        by_class.setdefault(cid, {"class_name": r["class_name"], "course_code": r["course_code"], "count": 0})
        by_class[cid]["count"] += 1

    # Last 7 days
    from datetime import timedelta
    today = date.today()
    weekly = []
    for i in range(6, -1, -1):
        d = (today - timedelta(days=i)).isoformat()
        cnt = sum(1 for r in records if r["date"] == d)
        weekly.append({"date": d, "count": cnt})

    return {
        "enrolled_classes": enrolled_count,
        "total_sessions": total_sessions,
        "by_class": list(by_class.values()),
        "weekly": weekly,
    }


@router.get("/reports/{student_id}")
async def reports(student_id: str, period: str = "month", _=Depends(get_current_user)):
    """period: week | month | all"""
    db = get_db()
    from datetime import timedelta

    today = date.today()
    if period == "week":
        cutoff = (today - timedelta(days=7)).isoformat()
    elif period == "month":
        cutoff = (today - timedelta(days=30)).isoformat()
    else:
        cutoff = "2000-01-01"

    records = await db.attendance.find({
        "student_id": student_id,
        "date": {"$gte": cutoff},
    }).sort("date", -1).to_list(500)

    return [_fmt(r) for r in records]


@router.get("/class-report/{class_id}")
async def class_report(class_id: str, date_str: str = None, _=Depends(get_current_user)):
    """All student attendance for a class on a given date (default today)."""
    db = get_db()
    target_date = date_str or date.today().isoformat()
    records = await db.attendance.find({
        "class_id": class_id,
        "date": target_date,
    }).to_list(200)
    return [_fmt(r) for r in records]
